exporter/elasticsearchexporter/elasticsearch_bulk.go

```diff
@@ -182,70 +171,74 @@ func newBulkIndexer(logger *zap.Logger, client *elasticsearch7.Client, config *C
 		// See https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32344
 		maxDocRetry = config.Retry.MaxRequests - 1
 	}
-
-	pool := &bulkIndexerPool{
-		wg:    sync.WaitGroup{},
-		items: make(chan esBulkIndexerItem, config.NumWorkers),
-		stats: bulkIndexerStats{},
+	manager.pool = &sync.Pool{
+		New: func() any {
+			bi, err := docappender.NewBulkIndexer(docappender.BulkIndexerConfig{
+				Client:                client,
+				MaxDocumentRetries:    maxDocRetry,
+				Pipeline:              config.Pipeline,
+				RetryOnDocumentStatus: config.Retry.RetryOnStatus,
+			})
+			if err != nil {
+				return fmt.Errorf("error creating docappender bulk indexer: %w", err)
+			}
+			return bi
+		},
 	}
-	pool.wg.Add(numWorkers)
-
-	for i := 0; i < numWorkers; i++ {
-		bi, err := docappender.NewBulkIndexer(docappender.BulkIndexerConfig{
-			Client:                client,
-			MaxDocumentRetries:    maxDocRetry,
-			Pipeline:              config.Pipeline,
-			RetryOnDocumentStatus: config.Retry.RetryOnStatus,
-		})
-		if err != nil {
-			return nil, err
-		}
-		w := worker{
-			indexer:       bi,
-			items:         pool.items,
-			flushInterval: flushInterval,
-			flushTimeout:  config.Timeout,
-			flushBytes:    flushBytes,
-			logger:        logger,
-			stats:         &pool.stats,
-		}
-		go func() {
-			defer pool.wg.Done()
-			w.run()
-		}()
+
+	// Create a bulk indexer once to validate the config options
+	if err, ok := manager.pool.Get().(error); ok {
+		return nil, err
 	}
-	return pool, nil
+
+	return manager, nil
 }
 
 type bulkIndexerStats struct {
 	docsIndexed atomic.Int64
 }
 
-type bulkIndexerPool struct {
-	items chan esBulkIndexerItem
-	wg    sync.WaitGroup
-	stats bulkIndexerStats
+type bulkIndexerManager struct {
+	closeCh chan struct{}
+	stats   bulkIndexerStats
+	logger  *zap.Logger
+	config  *Config
+	wg      *sync.WaitGroup
+	sem     *semaphore.Weighted
+	pool    *sync.Pool
 }
 
-// Add adds an item to the bulk indexer pool.
-//
-// Adding an item after a call to Close() will panic.
-func (p *bulkIndexerPool) Add(ctx context.Context, index string, document io.WriterTo) error {
-	item := esBulkIndexerItem{
-		Index: index,
-		Body:  document,
+func (p *bulkIndexerManager) AddBatchAndFlush(ctx context.Context, batch []esBulkIndexerItem) error {
+	p.wg.Add(1)
+	defer p.wg.Done()
+
+	if err := p.sem.Acquire(ctx, 1); err != nil {
+		return err
 	}
-	select {
-	case <-ctx.Done():
-		return ctx.Err()
-	case p.items <- item:
-		return nil
+	defer p.sem.Release(1)
+
+	bi := p.pool.Get().(*docappender.BulkIndexer)
+	// Bulk indexer buffer should never contain any items, but double check for safety.
+	for bi.Items() != 0 {
+		p.logger.Error("bug: bulk indexer buffer contains unexpected residue")
+		bi = p.pool.Get().(*docappender.BulkIndexer)
+	}
+	defer p.pool.Put(bi)
```

Comment:
bi.Reset doesn't really do anything other than resetting stats, which it will on next request anyway. See [code](https://github.com/elastic/go-docappender/blob/main/bulk_indexer.go#L231). Do you think this should be considered a bug on docappender's end?

Reply:
I see, thanks, I don't think it's a bug.